File: .claude/hooks/guard_paths.py
#!/usr/bin/env python3
"""PreToolUse: bloquea escrituras fuera de las rutas permitidas del agente.

Lee el JSON del hook por stdin. Bloquea con exit 2 y el motivo en stderr.
Las rutas permitidas llegan en FACTORY_WRITE_PATHS (JSON) y el repo en FACTORY_REPO.
Fuera de una ejecución de la factoría (sin FACTORY_AGENT) no bloquea nada.
"""
import json
import os
import re
import sys
from pathlib import Path

WRITE_TOOLS = {"Write", "Edit", "MultiEdit", "NotebookEdit"}

# Nunca, ni siquiera dentro de write_paths: configuración de la factoría y secretos.
PROTECTED = ("factory/", ".claude/", ".git/", ".secrets/", "ops/runs/")

BANNED_BASH = [
    (r"\bgit\s+push\b", "git push está reservado al orquestador"),
    (r"\bgit\s+(checkout|switch)\s+main\b", "no se trabaja directamente sobre main"),
    (r"\brm\s+-rf\b", "borrado recursivo no permitido"),
    (r"\bsudo\b", "sudo no permitido"),
    (r"\.secrets\b", "acceso a secretos no permitido"),
    (r"\bcurl\b[^|]*\|\s*(ba)?sh\b", "descargar y ejecutar no permitido"),
]


def deny(reason: str) -> None:
    print(f"[guard_paths] bloqueado: {reason}", file=sys.stderr)
    sys.exit(2)


def relative(repo: Path, target: str) -> str | None:
    try:
        p = Path(target)
        p = p if p.is_absolute() else repo / p
        return str(p.resolve().relative_to(repo.resolve()))
    except (ValueError, OSError):
        return None  # fuera del repo


def main() -> None:
    agent = os.environ.get("FACTORY_AGENT")
    if not agent:
        sys.exit(0)  # uso interactivo: no aplica

    try:
        event = json.load(sys.stdin)
    except json.JSONDecodeError:
        sys.exit(0)

    repo = Path(os.environ.get("FACTORY_REPO", "."))
    allowed = json.loads(os.environ.get("FACTORY_WRITE_PATHS", "[]"))
    tool = event.get("tool_name", "")
    tool_input = event.get("tool_input") or {}

    if tool in WRITE_TOOLS:
        target = tool_input.get("file_path") or tool_input.get("notebook_path") or ""
        rel = relative(repo, target)
        if rel is None:
            deny(f"{agent} intenta escribir fuera del repositorio: {target}")
        if any(rel.startswith(p) for p in PROTECTED):
            deny(f"{rel} es una ruta protegida")
        if not any(rel.startswith(p) for p in allowed):
            deny(f"{agent} solo puede escribir en {allowed}; intentaba {rel}")

    elif tool == "Bash":
        command = tool_input.get("command", "")
        for pattern, reason in BANNED_BASH:
            if re.search(pattern, command):
                deny(reason)

    sys.exit(0)


if __name__ == "__main__":
    main()

File: .claude/hooks/guard_paths.ts
/**
 * PreToolUse: bloquea escrituras fuera de las rutas permitidas del agente.
 *
 * Lee el JSON del hook por stdin. Bloquea con exit 2 y el motivo en stderr.
 * Las rutas permitidas llegan en FACTORY_WRITE_PATHS (JSON) y el repo en FACTORY_REPO.
 * Fuera de una ejecución de la factoría (sin FACTORY_AGENT) no bloquea nada.
 * Dentro de una ejecución falla cerrado: cualquier error interno bloquea, porque para
 * Claude Code un hook que termina con otro código distinto de 2 no bloquea.
 */
import { readFileSync, realpathSync } from "node:fs";
import { basename, dirname, isAbsolute, join, relative, resolve } from "node:path";

const WRITE_TOOLS = new Set(["Write", "Edit", "MultiEdit", "NotebookEdit"]);

// Nunca, ni siquiera dentro de write_paths: configuración de la factoría y secretos.
const PROTECTED = ["factory/", ".claude/", ".git/", ".secrets/", "ops/runs/"];

const BANNED_BASH: [RegExp, string][] = [
  [/\bgit\s+push\b/, "git push está reservado al orquestador"],
  [/\bgit\s+(checkout|switch)\s+main\b/, "no se trabaja directamente sobre main"],
  [/\brm\s+-rf\b/, "borrado recursivo no permitido"],
  [/\bsudo\b/, "sudo no permitido"],
  [/\.secrets\b/, "acceso a secretos no permitido"],
  [/\bcurl\b[^|]*\|\s*(ba)?sh\b/, "descargar y ejecutar no permitido"],
  // Una dependencia nueva se declara en package.json y se revisa en la PR; `npm install` a
  // secas (sin paquete) sí se permite, es el que instala lo ya declarado.
  // Los flags se saltan: lo que se busca es un paquete suelto detrás del subcomando.
  [/\bnpm\s+(?:i|install|add)\b(?:\s+-{1,2}[\w-]+)*\s+(?!-)\S/,
    "declara la dependencia en package.json y usa `npm install` sin argumentos"],
  [/\bnpx\b|\bnpm\s+(?:create|init|exec)\b/, "ejecutar un paquete que no está en package.json no está permitido"],
  // Revisar es informar: quien decide si una PR entra es una persona.
  [/\bgh\s+pr\s+(?:merge|review|close|ready)\b/, "aprobar, cerrar o fusionar una PR no es de un agente"],
];

function deny(reason: string): never {
  process.stderr.write(`[guard_paths] bloqueado: ${reason}\n`);
  process.exit(2);
}

/** realpath que admite rutas que aún no existen: resuelve el antecesor que sí existe. */
function real(p: string): string {
  try {
    return realpathSync(p);
  } catch {
    const parent = dirname(p);
    return parent === p ? p : join(real(parent), basename(p));
  }
}

/** Ruta relativa al repo, o null si queda fuera. */
function relativeToRepo(repo: string, target: string): string | null {
  const root = real(repo);
  const rel = relative(root, real(isAbsolute(target) ? target : resolve(repo, target)));
  return rel.startsWith("..") || isAbsolute(rel) ? null : rel;
}

function main(agent: string): void {
  const event = JSON.parse(readFileSync(0, "utf8"));
  const repo = process.env.FACTORY_REPO || ".";
  const allowed: string[] = JSON.parse(process.env.FACTORY_WRITE_PATHS || "[]");
  const tool: string = event.tool_name ?? "";
  const input = event.tool_input ?? {};

  if (WRITE_TOOLS.has(tool)) {
    const target: string = input.file_path || input.notebook_path || "";
    const rel = relativeToRepo(repo, target);
    if (rel === null) deny(`${agent} intenta escribir fuera del repositorio: ${target}`);
    if (PROTECTED.some((p) => rel.startsWith(p))) deny(`${rel} es una ruta protegida`);
    if (!allowed.some((p) => rel.startsWith(p))) {
      deny(`${agent} solo puede escribir en ${JSON.stringify(allowed)}; intentaba ${rel}`);
    }
  } else if (tool === "Bash") {
    const command: string = input.command ?? "";
    for (const [pattern, reason] of BANNED_BASH) if (pattern.test(command)) deny(reason);
  }
}

const agent = process.env.FACTORY_AGENT;
if (agent) {
  try {
    main(agent);
  } catch (e) {
    deny(`error interno del guardián: ${(e as Error).message}`);
  }
}
